### PR TITLE
esp32/machine_uart: Add uart_mode/flow_ctrl args to UART construct/init.

### PR DESCRIPTION
**Pull request by gmasse**
Give access to hardware flow control and RS485 half-duplex mode.

### CODE REVIEW

**Comment by dmartauz** [comment]
@gmasse The idea is that DE/RE# pins of RS485 transciever are controlled by UART peripheral of ESP32 and not by the application code? If so, can any output capable GPIO be used for this signal?

**Comment by gmasse** [comment]
@dmartauz In some way yes. ESP32 UART does not directly support half-duplex RS485 but the ESP-IDF software implementation can automatically command ~RE/DE via the RTS pin (details in [ESP-IDF documentation](https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/peripherals/uart.html#overview-of-rs485-specific-communication-options)).

So if you want to command ~RE/DE from ESP32, you have to disable hardware flow control (`flow_ctrl=0`) and change mode to RS485 half-duplex (`uart_mode=1`). You can also choose any GPIO output for RTS pin (`rts` argument).

**Comment by gmasse** [comment]
Allowed values for `flow_ctrl` are: (based on ESP-IDF [uart_hw_flowcontrol_t](https://github.com/espressif/esp-idf/blob/143d26aa49df524e10fb8e41a71d12e731b9b71d/components/driver/include/driver/uart.h#L103) enum)
    0 --> disable hardware flow control
    1 --> enable RX hardware flow control (rts)
    2 --> enable TX hardware flow control (cts)
    3 --> enable hardware flow control

Allowed values for `uart_mode` are: (based on ESP-IDF [uart_mode_t](https://github.com/espressif/esp-idf/blob/143d26aa49df524e10fb8e41a71d12e731b9b71d/components/driver/include/driver/uart.h#L52) enum)
    0 --> regular UART mode
    1 --> half duplex RS485 UART mode control by RTS pin
    2 --> IRDA UART mode
    3 --> RS485 collision detection UART mode (used for test purposes)
    4 --> application control RS485 UART mode (used for test purposes)

File: ports/esp32/machine_uart.c [hunk 1]
```diff
@@ -48,12 +48,17 @@ typedef struct _machine_uart_obj_t {
...
+    uint16_t timeout_char;        // timeout waiting between chars (in ms)
+    uint32_t invert;              // lines to invert
+    uint8_t uart_mode;            // UART mode
+    uint8_t flow_ctrl;            // hardware flow control
+    uint8_t rx_flow_ctrl_thresh;  // hardware flow control RTS threshold
 } machine_uart_obj_t;
 
 STATIC const char *_parity_name[] = {"None", "1", "0"};
+STATIC const char *_uart_mode_name[] = {"UART", "RS485_HALF_DUPLEX", "IRDA", "RS485_COLLISION_DETECT", "RS485_APP_CTRL"};
+STATIC const char *_flow_ctrl_name[] = {"DISABLE", "RTS", "CTS", "CTS_RTS"};
```
Comment (tve) [inline ports/esp32/machine_uart.c]: These should be python constants on `machine.UART` IMHO.

File: ports/esp32/machine_uart.c [hunk 2]
```diff
@@ -243,6 +251,59 @@ STATIC void machine_uart_init_helper(machine_uart_obj_t *self, size_t n_args, co
     }
     self->invert = args[ARG_invert].u_int;
     uart_set_line_inverse(self->uart_num, self->invert);
+
+    // set mode
+    switch (args[ARG_uart_mode].u_int) {
```
Comment (tve) [inline ports/esp32/machine_uart.c]: This switch should be eliminated by providing `machine.UART.xxx` constants that have the right value to pass to uart_set_mode and assignment to `uart_mode`.

File: ports/esp32/machine_uart.c [hunk 3]
```diff
@@ -243,6 +251,59 @@ STATIC void machine_uart_init_helper(machine_uart_obj_t *self, size_t n_args, co
...
+        default:
+            mp_raise_ValueError("invalid UART mode");
+            break;
+    }
+
+    // set hardware flow control threshold (RTS)
+    self->rx_flow_ctrl_thresh = args[ARG_rx_flow_ctrl_thresh].u_int;
+
+    // set hardware flow control
+    switch (args[ARG_flow_ctrl].u_int) {
```
Comment (tve) [inline ports/esp32/machine_uart.c]: Eliminate switch by choosing constants of the appropriate values.

**Review by tve — COMMENTED** [review]
IMHO:
- this PR requires documentation of the new KW args in the machine.uart class, with a note about being esp32-only for now
- the uart_mode and flow_ctrl args need to be range-checked somewhere
- I believe flow_ctrl is generic (can be implemented using the same interface on other micros) although I wonder whether there should be some generic values for rx_flow_ctrl_thres (maybe percentages of available buffer), or perhaps at least the default value should be a "decent" default (I don't believe that 0 is a reasonable default).
- uart_mode is a bit esp32-specific, although other uC's also have rs485 and IrDA support, dunno whether the API is "portable" given appropriate values for this parameter

File: ports/esp32/machine_uart.c (hunk 1 above)
Comment (gmasse) [reply ports/esp32/machine_uart.c]: Thanks for comment @tve.
I am not expert of Micropython. Any pointer to help me implementing a python constant?

Comment (tve) [reply ports/esp32/machine_uart.c]: For example https://github.com/micropython/micropython/blob/master/ports/esp32/machine_uart.c#L383-L386

**Comment by mrvelic** [comment]
@gmasse We'd really like to be able to use this functionality without a manual patch as we have a RS485 driver that requires this configuration. Are you at all interested in addressing the comments?
I am happy to attempt to address them myself, but unsure of the etiquette around doing so.

**Comment by gmasse** [comment]
@mrvelic I do not have time to fix all the comments. It would be great if you can help !

**Comment by mrvelic** [comment]
@gmasse No worries, @tve do you guys have some sort of process for co-contributing or taking over a PR?

**Comment by mrvelic** [comment]
@gmasse @tve I have updated my fork based on this PR here: https://github.com/mrvelic/micropython/commit/e6a8481e31d64845b4774ecc2f5b973dacb90351
Let me know what the correct way to proceed is, I have also built and flashed an esp32 and tested various mode combinations.

**Comment by mrvelic** [comment]
@tve just wondering what the process is for taking over the PR again, should I just make a new one and this one will be declined?

**Comment by dpgeorge** [comment]
This was implemented in #6902 (see a3675294ae39a724aec6959238c45af0dea92f21) via the flow keyword argument, which matches stm32 and rp2.

**Pull request closed without merging**